[css-animations] Fix serialization and parsing of `animation-name` strings

nt1m opened the pull request:
#### 82d81077a049566638c9bb6a3022e2621d69934a
<pre>
[css-animations] Fix serialization and parsing of `animation-name` strings
<a href="https://bugs.webkit.org/show_bug.cgi?id=285347">https://bugs.webkit.org/show_bug.cgi?id=285347</a>
<a href="https://rdar.apple.com/142318879">rdar://142318879</a>

Reviewed by Matthieu Dubet.

<a href="https://drafts.csswg.org/css-animations/#typedef-keyframes-name">https://drafts.csswg.org/css-animations/#typedef-keyframes-name</a>

`&lt;string&gt;` values should be parsed and serialized as identifiers unless there is ambiguity (in case of `none` or CSS-wide keywords).

`&quot;none&quot;` as a string should be treated as an keyframe name, not the `none` keyword.

These changes match the spec and Firefox&apos;s behavior.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-name-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-name-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-name-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-name-valid.html:
Add more testcases.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/animation-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/animation-name.html:
Remove erroneous testcase that assumed strings were invalid (Chrome also fails it).

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.cpp:
(WebCore::CSSPropertyParserHelpers::consumeKeyframesName): Correctly handle string values and convert them to custom idents whenever possible.

Canonical link: <a href="https://commits.webkit.org/288404@main">https://commits.webkit.org/288404@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6e9fd331ceb7d6bf3569e5f7db752b60367e70b7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83123 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/2761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37418 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/88228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/34165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85215 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/2840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10689 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/88228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/34165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86178 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/2840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/75585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/88228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/2840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/29785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/33200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/2840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/30485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/89595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/10406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/89595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/10635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/71397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/89595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12845 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/10359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/15860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/10227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/13693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/11996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-name-computed-expected.txt:
```diff
@@ -1,10 +1,29 @@
 PASS Property animation-name value 'paused'
 PASS Property animation-name value 'first, second, third'
-PASS Property animation-name value '"something"'
+PASS Property animation-name value '"string"'
```
Maybe `"foobar"` or `"something"` would be clearer ?

mdubet commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -608,9 +608,9 @@ void ViewTransition::setupDynamicStyleSheet(const AtomString& name, const Captur
     {
         CSSValueListBuilder list;
-        list.append(CSSPrimitiveValue::create("-ua-view-transition-fade-out"_s));
+        list.append(CSSPrimitiveValue::createCustomIdent("-ua-view-transition-fade-out"_s));
```
Could those changes be in a separate patch ?

nt1m committed:
[css-animations] Fix serialization and parsing of `animation-name` strings
https://bugs.webkit.org/show_bug.cgi?id=285347
rdar://142318879

Reviewed by Matthieu Dubet.

https://drafts.csswg.org/css-animations/#typedef-keyframes-name

`<string>` values should be parsed and serialized as identifiers unless there is ambiguity (in case of `none` or CSS-wide keywords).

`"none"` as a string should be treated as an keyframe name, not the `none` keyword.

These changes match the spec and Firefox's behavior.

* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-name-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-name-computed.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-name-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/parsing/animation-name-valid.html:
Add more testcases.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/animation-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/animation-name.html:
Remove erroneous testcase that assumed strings were invalid (Chrome also fails it).

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.cpp:
(WebCore::CSSPropertyParserHelpers::consumeKeyframesName): Correctly handle string values and convert them to custom idents whenever possible.

Canonical link: https://commits.webkit.org/288404@main

weinig commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.cpp:
```diff
@@ -42,8 +42,9 @@ RefPtr<CSSValue> consumeKeyframesName(CSSParserTokenRange& range, const CSSParse
-        if (equalLettersIgnoringASCIICase(token.value(), "none"_s))
-            return CSSPrimitiveValue::create(CSSValueNone);
+        auto valueId = cssValueKeywordID(token.value());
+        if (isValidCustomIdentifier(valueId) && valueId != CSSValueNone)
```
Would you mind adding a comment explaining that there is prose indicating that "none" must be excluded. Somethings like:

```c++
// Per spec, "The <custom-ident> additionally excludes the `none` keyword."
```

Not necessary for this change, but perhaps adding a `isValidCustomIdentifierExcluding<CSValueID... IdsToExclude>(CSSValueID)` would be a good helper in CSSParserIdioms.h.

I have been meaning to suggest to the CSSWG, and this reminds me I should go do that, an extension to the grammar syntax to allow additional exclusions. Something along the lines of <custom-ident excluding="none, foo"> to make these rules more clear.
